lib/std/os/windows.zig

```diff
@@ -539,7 +564,16 @@ pub fn WriteFile(
         loop.beginOneEvent();
         suspend {
             const adjusted_len = math.cast(DWORD, bytes.len) catch maxInt(DWORD);
-            _ = kernel32.WriteFile(handle, bytes.ptr, adjusted_len, null, &resume_node.base.overlapped);
+            if (kernel32.WriteFile(handle, bytes.ptr, adjusted_len, null, &resume_node.base.overlapped) == FALSE) {
+                switch (kernel32.GetLastError()) {
+                    .IO_PENDING => {},
+                    else => |err| {
+                        resume @frame();
```

Comment:
Something along these lines?
```zig
    var acceptErr: ?ws2_32.WinsockError = null;
    var eventCompletedOutsideLoop = false;

    suspend {
        if (ws2_32.AcceptEx(
            listening_s,
            accepting_s,
            &buffer,
            0,
            addressSize + 16,
            addressSize + 16,
            &syncBytesRead,
            &resume_node.base.overlapped,
        ) == FALSE) {
            switch (ws2_32.WSAGetLastError()) {
                .WSA_IO_PENDING => {},
                else => |err| {
                    acceptErr = err;
                    eventCompletedOutsideLoop = true;
                    resume @frame();
                },
            }
        } else {
            //Completed synchroniously
            eventCompletedOutsideLoop = true;
            resume @frame();
        }
    }

    if(eventCompletedOutsideLoop)
    {
        loop.finishOneEvent();
    }

    if (acceptErr) |err| {
        return err;
    }
```

It's a shame that `loop.finishOneEvent` can't be `defer`'ed
I splitted the logic of error handling and finishing event as either way two variables would be needed, or I could place `loop.finishOneEvent` before each resume

Reply:
Yea, this is what id imagine to do. Also you can make the loop event calls unconditional regardless of the suspend:
```rs
loop.beginOneEvent();
defer loop.finishOneEvent();

suspend {
...
}
```

Reply:
If only it would be that simple
Code sample from loop.zig : 1448
```zig
                .windows => {
                    var completion_key: usize = undefined;
                    const overlapped = while (true) {
                        var nbytes: windows.DWORD = undefined;
                        var overlapped: ?*windows.OVERLAPPED = undefined;
                        switch (windows.GetQueuedCompletionStatus(self.os_data.io_port, &nbytes, &completion_key, &overlapped, windows.INFINITE)) {
                            .Aborted => return,
                            .Normal => {},
                            .EOF => {},
                            .Cancelled => continue,
                        }
                        if (overlapped) |o| break o;
                    } else unreachable; // TODO else unreachable should not be necessary
                    //Rest of the code ...
                    resume handle;
                    self.finishOneEvent(); // <--- Here
                },
```

The worker thread finishes event after completion. I'm not sure why this design choice was made, but I don't wanna change it without knowing full context

Reply:
Oh, you're right.

TFW #8224